src/mesh_tools.rs

```diff
@@ -0,0 +1,279 @@
+use std::collections::HashMap;
+
+use smallvec::SmallVec;
+
+use crate::convert::{cast_u32, cast_usize};
+use crate::geometry::Geometry;
+use crate::mesh_topology_analysis::face_to_face_topology;
+
+fn crawl_faces(
+    current_face_index: u32,
+    face_to_face: &HashMap<u32, SmallVec<[u32; 8]>>,
+    available_faces: &mut Vec<u32>,
+    separate_faces: &mut Vec<u32>,
+) {
+    separate_faces.push(current_face_index);
+
+    if let Some(all_neighbors) = face_to_face.get(&current_face_index) {
+        let neighbors: Vec<u32> = all_neighbors
+            .iter()
+            .copied()
+            .filter(|n| available_faces.iter().any(|a_f| a_f == n))
+            .collect();
+        let af: Vec<_> = available_faces
+            .iter()
+            .copied()
+            .filter(|f| neighbors.iter().all(|n| n != f))
+            .collect();
+        *available_faces = af;
+        for neighbor_index in neighbors {
+            crawl_faces(
+                neighbor_index,
+                face_to_face,
+                available_faces,
+                separate_faces,
+            );
+        }
+    }
+}
+
+/// Crawls the geometry to find continuous patches of geometry.
+/// Returns a vector of new separated geometries.
+#[allow(dead_code)]
+pub fn separate_isolated_meshes(geometry: &Geometry) -> Vec<Geometry> {
+    let face_to_face = face_to_face_topology(geometry);
+    let faces = geometry.faces();
+    let vertices: Vec<_> = geometry.vertices().to_vec();
+    let normals: Vec<_> = geometry.normals().to_vec();
+
+    let mut available_faces: Vec<u32> = face_to_face.keys().map(|key| cast_u32(*key)).collect();
```

Comment:
I'd name both this and separate_faces: `available_face_indices` and `separate_face_indices`.